### PR TITLE
Reduce memory footprint on export and import to avoid OOM issues.

### PR DESCRIPTION
**Pull request by mrimer**
[Relevant issue](https://forum.caravelgames.com/viewtopic.php?TopicID=38449)

This PR addresses OOM errors when attempting to export and import very large .player files (and possibly also .hold files, etc).

Here's what's in this change, which I welcome your feedback on if you have thoughts/opinions:
* Up till now, when exporting, all the data being compiled into the file is first concatenated into a text string in RAM, which is then zlib-compressed into another buffer in RAM, and then CStretchyBuffer-encoded, then written to a file on disk.  Large player profiles (or maybe holds) can run out of memory (in the 32-bit address space in our Windows build) during export.
* To mitigate this, I've changed the export process to use gzip verbs (instead of zlib format), which allow writing of multiple compressed data segments to a file on disk.  I've inserted this operation into the callbackf() method that's called periodically.  Whenever the string gets large (10MB is the heuristic threshold I chose), the buffer is written to disk and cleared out.  Multiple gzip-compressed chunks may be written in sequence to the file.  Now memory usage stays low during export!
* I removed the "Encode" call, which I'm not sure we really need (it's not really secure anyway).  Note this change enables treating the file written to disk as a standard gzip file, and the XML contents can be viewed by anyone who wishes to (rename it as a .gz file and) unpack the contents.  This is probably not a bad thing, but we could re-introduce the encode/decode step if we really want to.
* Both gzip and zlib-plus-encode formats are still supported for import.  Only the gzip version will be written going forward for exports.

_This format change would require adding code to the server to process new v5.1.1+ compressed file uploads, to check for a gzip-format file header (see DbXML.cpp changes) and uncompress the file from gzip format instead of zlib+encode format in that case._   (You don't need to check for any versioning info; just the file header for gzip id bytes.)

The above changes don't address import memory issues.  I spent a good while looking at the DROD address space, which appears to be highly fragmented for large .dat files (by Metakit, if I'm concluding correctly).  I tried out using a custom memory allocator, which greatly reduces the number of distinct mallocs that take place across the address space for sets and maps.  This helped reduce # of allocations (except Metakit char[] allocations).  In certain cases w/ large .dats, I found I wasn't even able to malloc 80MB of contiguous address space, even though the app itself is only using 100-200MB of RAM.  This is ridiculous, but without swapping out Metakit (which would be a lot of work), very large imports would need to be addressed in another way.  If you have any ideas, feel free to share them.  (I can add the custom stl container memory allocator implementation I used in a future PR if it seems interesting to you; I've not included it here.)

Edit: I made a heuristic update to reduce potential import memory allocation, which fortunately resolves the OOM issue on uncompression with the large .player files and large .dat files I tested with.  (Yay!)

Edit2: I also added a progress indicator to the status text being displayed while demos and saved games are being validated for each hold.

### CODE REVIEW

File: DRODLib/DbXML.cpp
Comment (EvidentlyCube) [inline DRODLib/DbXML.cpp:44]: Can you add a comment explaining what it is?

File: DRODLib/DbXML.cpp
Comment (EvidentlyCube) [inline DRODLib/DbXML.cpp:248]: Extract the magic number to a constant at the top of the file perhaps?

**Review by EvidentlyCube — APPROVED** [review]
Just two minor things

**Comment by EvidentlyCube** [comment]
I am wondering - right now importing loads the whole file into the memory and then imports it. Is there a reason why it couldn't be solved by streaming the file and decoding/inflating on the fly whenever next bytes are requested?

File: DRODLib/DbXML.cpp
Comment (mrimer) [reply DRODLib/DbXML.cpp]: Sure. Done!

File: DRODLib/DbXML.cpp
Comment (mrimer) [reply DRODLib/DbXML.cpp]: Good idea. Thank you. Done!

**Comment by mrimer** [comment]
> I am wondering - right now importing loads the whole file into the memory and then imports it. Is there a reason why it couldn't be solved by streaming the file and decoding/inflating on the fly whenever next bytes are requested?

That's the right question to ask.  In theory, this is possible.  I believe it could require a large refactor of the XML parsing code, but maybe it's simpler than that.  It would start around CDbXML::ImportXML(), which performs two passes of the uncompressed data buffer.  The uncompression and the parsing steps would need to be coupled.

The first pass of the XML data scans through all the records and counts how many of each type there are, in order to have Metakit pre-allocate space for them to reduce memory address fragmentation.

The second pass reads, populates, identifies, updates foreign keys, and inserts all the records into the DB.

I think that it may be possible to feed in a part of the uncompressed data in a loop for each of these two passes.  If the parser throws an "incomplete data" error, then uncompress more data, append to whatever remained to parse, and continue parsing.

Hopefully this is unnecessary at this point, but if it is ever required to import a large file, this is probably the simplest method to accomplish it.

**Comment by mrimer** [comment]
Updated to address review comments.

**Comment by EvidentlyCube** [comment]
Looks good! While we're at it, can you copy-paste your description of how import works and put it somewhere in the code where someone trying to understand would stumble upon it? Or expand it if you think it'd be a good idea.

For "where", I think `DBXml.cpp` is a good place, and in each `Export` method in the `Db*` classes you could just add a small note "Export/Import is documented in DBXml.cpp". It would certainly help future maintainers :).

**Comment by mrimer** [comment]
@EvidentlyCube , great idea about documentation.  Definitely want future maintainers to be able to easily understand the code design and flow.

Please reference the core `DbXML::ImportXML` method definition for detailed documentation about the import process.

I'll add some more documentation, but generally export/import have a good amount of explanation already in place in DBXML.cpp.

At this point, I believe @schikore still needs to greenlight this general new approach to encoding the exported data, as it will impact server imports.  Once we get approval, we can merge this PR.

**Comment by mrimer** [comment]
Update: I believe I've devised a way to simplify piecemeal uncompression in the import process to further reduce risk of OOM issues.  I'll be adding more documentation and refactoring to make ImportXML easier to follow.  To avoid potential confusion here, I'll prepare that as a distinct PR that I'll put up once this one is approved and merged.

**Comment by schikore** [comment]
> At this point, I believe @schikore still needs to greenlight this general new approach to encoding the exported data as gzip instead of zlib+encode, as it will impact server imports. Once we get approval, we can merge this PR.

@mrimer Can you provide me a file compressed with this new code so that I can run a quick test on the server to make sure it's fine?  I believe it will be, but would rather check to be  sure.

**Comment by schikore** [comment]
Okay, I've got code on the forum that will determine which format is used and properly decode it.  Looks good to me.

**Review by schikore — APPROVED** [review]

**Pull request merged**
